---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)

baseDir = "/home/mistrm82/R/sharpePD1_rnaseq"
project_summary = file.path(baseDir, "project-summary.csv")
counts_file = file.path(baseDir, "data/combined.counts")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]

# Load metadata
meta = read.csv(file.path(baseDir, "meta/sharpePD1_project.csv"), row.names=1)
meta = meta[order(rownames(meta)),]
rows <- sapply(row.names(meta), function(x){strsplit(x, split=".", fixed=T)[[1]][1]})
row.names(meta) <- unname(rows)


# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")
```

```{r heatmap-function, echo=FALSE}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads

The number of mapped reads are quite low across the board (maximum ~ 6 millions reads). There are five samples that are on the very low range (less than 2 million reads), and these samples correspond to the Cre- PD1- sample group (the 6th sample from this group is also quite low with just over 2million reads). The Cre+ PD1+ samples fall in the higher range of mapped reads; which are the samples we are most interested in for our comparison.

```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("") +
    theme(axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25))) 
```

## Genomic mapping rate
A good mapping rate across the board. 

```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    geom_bar(stat="identity") +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1),
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25)))
```

## Unique mapping rate

Although ~75% of the total reads in each sample are mapping, less than 40% of those are uniquely mapped (multi-mappers are ignored when counting). The sequence duplication plots from the fastqc reports show high levels of s

```{r unique-rate-plot}
dd = data.frame(Name=names(counts), Unique=colSums(counts), Mapped=summarydata[,"Mapped"])
ggplot(dd, aes(x=Name, y=Unique/Mapped)) +
    geom_bar(stat="identity") +
    ylab("unique mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25)))
```

## Number of genes detected

Usually, for mouse we observe ~15k genes to be detected - here we see half that (and in some samples, less). This is in line with the low exonic mapping rate that we observe (see below).

```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25))) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
```{r saturation-plot}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25))) +
    ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate

The next two figures gives us insight as to where reads are mapping. There are an unusually low number of reads mapping to exonic regions (< 50%) and higher than normally observed mapping to intergenic (~40%; see below). Usually this would mean possible genomic DNA contamination and/or pre-mRNA. 

```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25))) +
    ylab("exonic mapping rate") + xlab("")
```


## Intergenic mapping rate
```{r intergenic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Intergenic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25))) +
    ylab("intergenic mapping rate") + xlab("")
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=45, hjust=1), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25))) +
    ylab("rRNA rate") + xlab("")
```


## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90), 
          axis.title = element_text(size = rel(1.5)),
          axis.text = element_text(size = rel(1.25))) + 
  xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
The two outlier samples correspond to samples that have less than 1 million reads mapped.

```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## MDS plot of TMM-normalized counts
There is no clear segregation of samples by group.

```{r mds-normalized, echo=FALSE}
# Perform PCA (requires transformed data matrix)
pca_obj <- prcomp(t(normalized_counts))

# Create data frame for input to ggplot
df <- cbind(meta,  pca_obj$x[,c('PC1', 'PC2')])

# Plot PC1 against PC2
ggplot(df, aes(PC1, PC2, color = samplegroup)) + 
  geom_text(aes(PC1, PC2, label = samplegroup), size = 3, hjust=0.1, vjust=0.1) +
  scale_x_continuous(expand = c(0.3,  0.3)) +
  theme(axis.title = element_text(size = rel(1.1)),
          axis.text = element_text(size = rel(1.1)))
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(counts),decreasing=TRUE)[1:30]
heatmap_fn(counts[select,])
```
